rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+    *   Note that we can still provide an Arrow -> StructuredTensor adapter to
+        achieve interoperability.
+
+What this alternative does not change / address:
+
+*   We still need the TensorAdapter API to convert from StructuredTensor to
+    other TF feedables, unless we are willing to only offer StructuredTensors to
+    our end-users (which might be the case eventually, but not likely to happen
+    soon). So a good portion of this proposal will remain mostly unchanged.
+*   We still cannot expose file-based or record-based interface to end-users in
+    the TF trainer. As that is a direct result of I/O + parsing being abstracted
+    out.
+
+The disadvantages of this approach are:
+
+*   This tightly couples TF with TFX components, in the following ways:
+
+    *   Components will need TF to read the data in.
+    *   Components that analyze the data (e.g. TFDV) will operate against
+        StructuredTensors, and the easiest way to to conduct certain
+        computations (for example, slicing a StructuredTensor or computing the
+        mean) with StructuredTensors is through TF ops and their python
+        bindings.
+
+    Such a coupling is not in our favor because:
+
+    *   Some TFX components functionally do not require TF to work. For example,
+        TFDV can analyze any data set. TFMA can analyze a model while treating
+        the model as a blackbox. In both cases, the ML library that trained the
+        model is irrelevant and TF should not be assumed. And coupling with TF
+        does not only introduce a heavy dependency, but also forces the user to
+        learn about TF if they need to implement a TFXIO for their data format.
+
+    *   Operations against StructuredTensors boils down to Python + TF ops. The
+        overhead of either is much higher than just calling an Arrow C++ API
+        that does the same operation.
+
+    *   The main extension point of TF is Ops which don't understand the nested
+        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
+        deals with StructuredTensors will be much complicated.
+
+# Questions and Discussion Topics
+
+## OSS build / release issues
+
+### We don’t have a strong representation in the Arrow community
+
+This has led to some issues. For example, the PyPI/Wheel packaging for pyarrow
+currently is unfunded and lacks volunteers and there is a risk of support being
+dropped, but we do rely on the pyarrow wheel as TFX is released on PyPI.
+
+### ABI compatibility with libarrow
+
+The OSS library, tfx_bsl will depend on Arrow and TensorFlow’s DSOs (dynamic
+shared objects). Because both libraries currently expose C++ APIs, there are
+always risks of incompatible ABIs as TensorFlow and Arrow are likely to be built
+using different toolchains, we cannot completely eliminate the risks.
+
+With [Modular TensorFlow](https://github.com/tensorflow/community/pull/77),
+which replaced all the C++ APIs with C-APIs, we will be able to eliminate the
```

Comment:
Modular TensorFlow is not yet implemented. We should probably first check whether arrow and TF can coexist. The main problems are usually protobuf and absl.

Reply:
We're working on removing Boost from our build toolchain (though it may be required to build some unit tests). See https://issues.apache.org/jira/browse/ARROW-6637 for information on the project

Thrift is indeed required for reading Parquet files. We'd be glad to work with you to resolve any concerns around the build toolchain.

Ultimately migrating to Bazel from CMake might be a good thing long term, but that is a very large project.

Reply:
@wesm Thanks for the information!

Thrift is needed to generate parquet_types.h and parquet_types.cpp, but that could be a one-time effort, instead of installing bison and flex (dependencies) and build parquet_types.h/parquet_types.cpp in every compile. 

In tensorflow/io I added a patch with the following as a workaround:
https://github.com/tensorflow/io/blob/master/WORKSPACE#L89-L107

I will looking into Arrow repo, and will try to create PR if I find some place where I could help.

Reply:
Having the compiled files is not sufficient -- they depend on parts or the Thrift library (either libthrift.a or libthrift.so)

Reply:
@wesm Ah yes 👍 , I wasn't clear in the last comment. The thrift library will always be needed to build parquet support. Though having the complied files of parquet_types.[h|cpp] will drop the need to install `bison` and `flex` binary, which could be easier for other programs (e.g., tensorflow/io) to integrate Arrow C++.

Reply:
I see. I don't think we'd be opposed to checking those in.

Reply:
I opened https://issues.apache.org/jira/browse/ARROW-6821. Seems like a straightforward build toolchain improvement

Reply:
Currently the way tfx_bsl [1] (the proposed repo/package to host this project) and TFDV [2] build with arrow is that we build with headers shipped with the pyarrow wheel and link with libarrow.so and libarrow_python.so shipped with the same wheel. 

all we need is a way to communicate RecordBatches / Arrays with pyarrow. And there is indeed risk if we want to build with TF and pyarrow at the same time, because TF and pyarrow are not built with the same toolchain.

I believe @yongtang avoided the risk in tensorflow_io by using the IPC format between TF and pyarrow, and statically build and link with arrow's source in the arrow dataset op. However producing the IPC format AFAIK is not copy-free. I was hoping that the proposal of the ABI stable C-struct [3] can provide a safe way of passing things in-process.

[1] https://github.com/tensorflow/tfx-bsl
[2] https://github.com/tensorflow/data-validation
[3] https://github.com/apache/arrow/pull/5608

Reply:
Also, if we were to build the proposed tf.Dataset that does I/O and produces a DT_VARIANT tensor of Arrow RecordBatch, and the OP that translates a RecordBatch to Tensors, they will be statically linking against arrow source code instead of libarrow.so, because there we don't need to talk with pyarrow. (tensorflow_io is doing the same).